README.md

```diff
@@ -229,6 +243,253 @@ Benchee.run %{"Awesome stuff" => fn -> magic end }
+
+The defined function **must return whatever it wants following hooks and the benchmarking function to consider the input**. This means, that there usually are 3 kinds of before_scenarios:
+
+* you just want to invoke a side effect, in that case return the given input unchanged
+* you want to alter the given input, in that case alter the given input
+* you want to keep the given input but add some other data, in that case just return a tuple like `{original_input, new_fancy_data}`
+
+First the _global_ `before_scenario` is invoked, then the local one. This also means that the return value of the _global_ `before_scenario` is passed on into its _local_ sibling. That return value is then passed into `before_each` (if present) or the benchmarking function (if no `before_each` is present).
+
+Usage:
+
+```elixir
+Benchee.run %{
+  "foo" =>
+    {
+      fn(input) -> foo(input) end, # input is 3 here
+      before_scenario: fn(input) -> input + 1 end # input is 2 here
+    },
+  "bar" =>
+    fn(input) -> bar(input) end # input is 2 here (only global ran)
+}, inputs: %{
+  "input 1" => 1,
+  before_scenario: fn(input) -> input + 1 end # input is 1 here
+}
+```
+
+When might this be useful?
+* You want to start a process to be used in your scenario
+* Something that you want to use in the benchmark starts a process associated with the PID of `self()` (each scenario is executed in its own process so these can't be started before all)
+* Clearing the cache before a scenario
+
+#### before_each
+
+Is executed before each invocation of the benchmarking function/before every measurement. The input for `before_each` is either the `input` or what was returned by `before_scenario`. It works much like [`before_scenario`](#before_scenario), the only major difference being that it is invoked at another time. Most importantly, the function needs to return what the benchmarking function should receive as input.
+
+```elixir
+Benchee.run %{
+  "bar" => fn(input) -> bar(input) end # input will be 2 here
+}, inputs: %{
+  "input 1" => 1,
+  before_each: fn(input) -> input + 1 end
+}
+```
+
+When might this be useful?
+* Retrieving a record from the database and passing it on to the benchmarking function to do something(tm) without the retrieval from the database adding to the benchmark measurement
+* Busting caches so that all measurements are taken in an uncached state
+* Picking a random value from a collection and passing it to the benchmarking function for measuring performance with a wider spread of values
+
+### after_each
+
+Is executed directly after the invocation of the benchmarking function. It is passed the return value of the benchmarking function. Unlike the `before_*` functions the return value does not matter and isn't passed on. Both _global_ and _local_ `after_each` get the return value of the function.
+
```

Comment:
Important distinction between before/after. But then I'm confused. The return value isn't passed on, and then `after_each` gets the return value of the function. Seems like there are two different return values here: the hook return value and the benchmark return value?

Reply:
hm, the return value of `after_each` gets ignored and both local and global `after_each` get the return value of the "main" benchmarking function.

Passing on the `after_each` value didn't seem to make much sense - it could only get passed from "local" after_each to "global" after_each and that didn't seem to make much sense

Reply:
Do you think that still makes sense? If yes,  how can we make it easier to understang?